miniml/compiler/compile.scm

```diff
@@ -2233,6 +2252,48 @@
      (compile-letexits env stacksize exit-defs body))
     (('LExit exit args)
      (compile-exit env stacksize exit args))
+    (('LWhile cond body)
+     (let* ((labcond (newlabel))
+            (labloop (newlabel)))
+       (bytecode-BRANCH-to labcond)
+       (bytecode-emit-label labloop)
+       (bytecode-put-u32-le CHECKSIGNALS)
+       (compile-expr env stacksize body)
+       (bytecode-emit-label labcond)
+       (compile-expr env stacksize cond)
+       (bytecode-put-u32-le BRANCHIF)
+       (bytecode-emit-labref labloop)
+       (bytecode-CONSTINT 0)
+       ))
+    (('LFor v dir b1 b2 body)
+     (let* ((labloop (newlabel))
+            (labend (newlabel)))
+       (compile-expr env stacksize b1)
+       (bytecode-put-u32-le PUSH)
+       (compile-expr env (+ stacksize 1) b2)
+       (bytecode-put-u32-le PUSH)
+       (bytecode-put-u32-le PUSH)
```

Comment:
There would be less bytecode to read carefully if LFor used variables for its bound, with let-bindings introduced during lowering.

Reply:
This seems like a reasonable idea, but I'm unsure whether actually help: each access to the loop bounds would require a recursive call to get the variable, and I'm not sure how well it would interfere with the variable-subsitution optimization.

Reply:
One important advantage of having variables there is that, at the loop site, they do not need to consume stack space. We would still have a slack slot for the mutable bound variable, but this is strictly less stack-size-offset stuff to think about. (If we are worried about using three slots instead of two, we could also only make the end-bound a variable, which makes some sort of sense as it is the only one that, morally, gets evaluated several times.)

If the end-bound is a constant, indeed it would get inlined, so we get slightly larger code in the bound-checking part. However we rarely write `for i = ... to <constant> do`, the main case where I see this possibly happening in practice is `for i = ... downto 0 do`, in which case we have a `CONST0` opcode that makes it as compact as a small-index stack access.

(If you were thinking of having the start-bound passed as a variable and then mutate this variable directly -- in your comment about interference with variable-substitution -- then that idea was evil and of course it should not be done this way. I was thinking of redeclaring a mutable variable/slot locally.)